File: 15.3sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	x := 0
	y := 1
	z := len(nums) - 1
	result := make([][]int, 0)
	for x < y && y < z {
		if x > 0 && nums[x-1] == nums[x] {
			x++
			y = x + 1
			z = len(nums) - 1
			continue
		}
		for y < z {
			if y > x+1 && nums[y-1] == nums[y] {
				y++
				continue
			}
			if z < len(nums)-1 && nums[z+1] == nums[z] {
				z--
				continue
			}
			if nums[x]+nums[y]+nums[z] == 0 {
				result = append(result, []int{nums[x], nums[y], nums[z]})
				y++
				z--
				continue
			} else if nums[x]+nums[y]+nums[z] < 0 {
				y++
				continue
			} else if nums[x]+nums[y]+nums[z] > 0 {
				z--
				continue
			}
		}
		x++
		y = x + 1
		z = len(nums) - 1
	}
	return result
}

File: 15.3sum.ts
function threeSum(nums: number[]): number[][] {
  const result: number[][] = [];
  nums = nums.sort((a, b) => a - b);
  let x: number = 0;
  let y: number = 1;
  let z: number = nums.length - 1;

  while (x < y && y < z) {
    if (x > 0 && nums[x] === nums[x - 1]) {
      x++;
      y = x + 1;
      z = nums.length - 1;
      continue;
    }
    while (y < z) {
      if (y > x + 1 && nums[y] === nums[y - 1]) {
        y++;
        continue;
      }
      if (z < nums.length - 1 && nums[z] === nums[z + 1]) {
        z--;
        continue;
      }
      if (nums[x] + nums[y] + nums[z] === 0) {
        result.push([nums[x], nums[y], nums[z]]);
        z--;
        y++;
      } else if (nums[x] + nums[y] + nums[z] > 0) {
        z--;
      } else {
        y++;
      }
    }
    x++;
    y = x + 1;
    z = nums.length - 1;
  }

  return result;
}
